### PR TITLE
Stubbing and stuff

### PR DESCRIPTION
**Pull request by needmoney90**
Made these changes on a branch, I want to merge them into master.

### CODE REVIEW

**Review by mark-antony1 — COMMENTED** [review]
overall looks good. finish implementation before pushing

File: merkato/merkato.py [hunk 1]
```diff
@@ -25,23 +25,45 @@ def __init__(self, configuration, coin, base, spread, bid_reserved_balance, ask_
         self.distribution_strategy = 1
         self.spread = spread # i.e '.15
         self.profit_margin = profit_margin
-        # Create ladders from the bid and ask bidget here
+
+        # Exchanges have a maximum number of orders every user can place. Due
+        # to this, every Merkato has a balance of coins that are not currently
```
Comment (mark-antony1) [inline merkato/merkato.py]: change balance to reserve

File: merkato/merkato.py [hunk 2]
```diff
@@ -25,23 +25,45 @@ def __init__(self, configuration, coin, base, spread, bid_reserved_balance, ask_
...
-            print('total_pair_balances', total_pair_balances)
+
+            self._debug(100, 'total_pair_balances: ', total_pair_balances)
+
             allocated_pair_balances = get_allocated_pair_balances(configuration['exchange'], base, coin)
-            check_reserve_balances(total_pair_balances, allocated_pair_balances, coin_reserve=ask_reserved_balance, base_reserve=bid_reserved_balance)
+            funds_available = check_reserve_balances(total_pair_balances, allocated_pair_balances, coin_reserve=ask_reserved_balance, base_reserve=bid_reserved_balance)
+
+            if not funds_available:
+                # Don't create the Merkato. Maybe should raise?
```
Comment (mark-antony1) [inline merkato/merkato.py]: this is done in check_reserve_balances it will error and kill

File: merkato/merkato.py [hunk 3]
```diff
@@ -54,16 +76,31 @@ def __init__(self, configuration, coin, base, spread, bid_reserved_balance, ask_
...
         self.initialized = True  # to avoid being updated before orders placed
 
+
+    def kill():
+        ''' TODO: Function comment
+        '''
+        # Cancel all orders
+        self.cancelrange(ONE_SATOSHI, ONE_BITCOIN) # Technically not all, but should be good enough
+
+        # Remove reserve references in DB
```
Comment (mark-antony1) [inline merkato/merkato.py]: shouldn't this remove the relevant merkato from the db and then cancel all orders?

File: merkato/merkato.py [hunk 4]
```diff
@@ -75,20 +112,69 @@ def rebalance_orders(self, new_txes):
...
         for tx in ordered_transactions:
 
+            # Do a check for whether this particular tx refers to a filled order
+            partial_fill = is_partial_fill(tx['id']) # todo unimplemented
+
             if tx['type'] == SELL:
                 # print('amount', type(tx['amount']), type(tx[PRICE])) # todo use debug
-                
+
+                if partial_fill:
```
Comment (mark-antony1) [inline merkato/merkato.py]: extract everything under partial_fill to a seperate merkato class function

File: merkato/merkato.py [hunk 5]
```diff
@@ -75,20 +112,69 @@ def rebalance_orders(self, new_txes):
...
+                    self.base_partials_balance += tx['amount']
+                    # todo: modify partials balance in db as well
+
+                    # Update the last orderId (actually the id of the transaction)
+                    update_merkato(self.mutex_UUID, LAST_ORDER, tx['orderId'])
+
+                    # 3. Skip everything else
+                    continue
+
+                if tx['id'] in filled_orders:
```
Comment (mark-antony1) [inline merkato/merkato.py]: i would extract this to a variable called matching_order_exists or something

File: merkato/merkato.py [hunk 6]
```diff
@@ -75,20 +112,69 @@ def rebalance_orders(self, new_txes):
...
+
+                    # Update the last orderId (actually the id of the transaction)
+                    update_merkato(self.mutex_UUID, LAST_ORDER, tx['orderId'])
+
+                    continue
+
+                # We need to place a matching order
+                # We want to get the total amount of that order
+
+                total_amount = get_total_amount(tx['id']) # todo unimplemented
```
Comment (mark-antony1) [inline merkato/merkato.py]: this needs to be implemented

File: merkato/merkato.py [hunk 7]
```diff
@@ -115,11 +201,42 @@ def rebalance_orders(self, new_txes):
...
+                if tx['id'] in filled_orders:
+                    # Matching order has already been placed
+                    base_filled_sum += tx['amount']
+                    continue
+
+                filled_orders.append(tx['id'])
+                # We need to place a matching order
+                # We want to get the total amount of that order
+
+                total_amount = get_total_amount(tx['id']) # todo unimplemented
```
Comment (mark-antony1) [inline merkato/merkato.py]: implement

File: merkato/merkato.py [hunk 8]
```diff
@@ -211,6 +334,7 @@ def distribute_bids(self, price, total_to_distribute, step=1.02):
 
         # 4. Store the remainder of total_to_distribute, as well as the final
         #    order placed in decaying_bid_ladder
+        # TODO
```
Comment (mark-antony1) [inline merkato/merkato.py]: does this need to be done for this feature?

File: merkato/merkato.py [hunk 9]
```diff
@@ -397,44 +524,63 @@ def modify_settings(self, settings):
         # replace old settings with new settings
         pass
 
+
     def add_reserve(self):
+        ''' TODO: Function comment
```
Comment (mark-antony1) [inline merkato/merkato.py:520]: does this need to be implemented for this feature?

File: merkato/utils/database_utils.py [hunk 10]
```diff
@@ -59,25 +74,41 @@ def update_merkato(exchange_pair, key, value):
         conn.close()
 
 
+def kill_merkato(exchange_pair):
```
Comment (mark-antony1) [inline merkato/utils/database_utils.py]: implement

File: merkato/utils/database_utils.py [hunk 11]
```diff
@@ -129,98 +172,151 @@ def get_all_exchanges():
...
         c.execute('''SELECT * FROM merkatos WHERE exchange = ?''', (exchange,))
         exchanges = c.fetchall()
         conn.commit()
         conn.close()
+
         return exchanges
 
 
+def remove_base_partial(UUID, amount)
+pass
```
Comment (mark-antony1) [inline merkato/utils/database_utils.py]: implement

File: merkato/utils/database_utils.py [hunk 12]
```diff
@@ -129,98 +172,151 @@ def get_all_exchanges():
...
+
+
+def remove_quote_partial(UUID, amount)
+pass
```
Comment (mark-antony1) [inline merkato/utils/database_utils.py]: implement

File: merkato/utils/database_utils.py [hunk 13]
```diff
@@ -129,98 +172,151 @@ def get_all_exchanges():
...
+
+
+# I realize this could be done with negative values. For readability's sake,
+# i'm implementing it so you never need to pass in a negative number.
+def add_base_partial(UUID, amount)
+pass
```
Comment (mark-antony1) [inline merkato/utils/database_utils.py]: implement

File: merkato/utils/database_utils.py [hunk 14]
```diff
@@ -129,98 +172,151 @@ def get_all_exchanges():
...
+
+
+def add_quote_partial(UUID, amount)
+pass
```
Comment (mark-antony1) [inline merkato/utils/database_utils.py]: implement

**Review by mark-antony1 — CHANGES_REQUESTED** [review]

File: merkato/merkato.py (hunk 2 above)
Comment (needmoney90) [reply merkato/merkato.py]: Do we really want to do that though?

File: merkato/merkato.py (hunk 3 above)
Comment (needmoney90) [reply merkato/merkato.py]: Due to how we're handling partials, we need to sum up all the partial fills of all active orders before cancelling them, and then remove that amount from the partials balance.

File: merkato/merkato.py (hunk 2 above)
Comment (mark-antony1) [reply merkato/merkato.py]: Yes, I think that is appropriate, at least for now.

File: merkato/merkato.py (hunk 3 above)
Comment (mark-antony1) [reply merkato/merkato.py]: Why? If all the orders are cancelled it shouldn't matter. No additional trades will be placed or execute and the merkato will be wiped.

Comment (needmoney90) [reply merkato/merkato.py]: right of course, you're correct

File: merkato/merkato.py (hunk 8 above)
Comment (needmoney90) [reply merkato/merkato.py]: No, it's being handled by the decaying ladder itself.

File: merkato/merkato.py (hunk 9 above)
Comment (needmoney90) [reply merkato/merkato.py:520]: How do we initially set reserve again?

Comment (mark-antony1) [reply merkato/merkato.py:520]: in the merkato init, reserve is inputted by the user and then used to distribute ladders

Comment (needmoney90) [reply merkato/merkato.py:520]: This isn't necessary (yet).

**Comment by needmoney90** [comment]
Closing this in favor of #224

**Pull request closed without merging**